torch/csrc/jit/passes/frozen_linear_folding.cpp

```diff
@@ -0,0 +1,134 @@
+#include <torch/csrc/jit/ir/constants.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/fold_linear_bn.h>
+#include <torch/csrc/jit/passes/frozen_linear_folding.h>
+#include <torch/csrc/jit/passes/utils/optimization_utils.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/ones_like.h>
+#include <ATen/ops/zeros_like.h>
+#endif
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+
+bool supportedLinearNode(Node* n) {
+  if (n->kind() == aten::linear) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool FoldFrozenLinearBatchnorm(Block* b) {
+  bool graph_modified = false;
+  bool isBatchNorm1d = false;
+  for (Node* n : b->nodes()) {
+    for (Block* block : n->blocks()) {
+      graph_modified |= FoldFrozenLinearBatchnorm(block);
+    }
+
+    if (n->kind() == prim::Constant &&
+        n->output()->type()->isSubtypeOf(*StringType::get()) &&
+        toIValue(n->output()) == "expected 2D or 3D input (got {}D input)") {
+      isBatchNorm1d = true;
+    }
+
+    if (n->kind() == aten::batch_norm &&
+        supportedLinearNode(n->inputs().at(0)->node())) {
+      auto linear = n->inputs().at(0)->node();
+      auto bn = n;
+
+      if (nonConstantParameters(linear) || nonConstantParameters(bn)) {
+        continue;
+      }
+
+      auto bn_rm_ivalue = bn->namedInput("running_mean");
+      auto bn_rv_ivalue = bn->namedInput("running_var");
+
+      // check running_mean and running_var has value, if they are
+      // None(track_running_stats=False), skiping the folding path.
+      if (bn_rm_ivalue->type() == NoneType::get() &&
+          bn_rv_ivalue->type() == NoneType::get()) {
+        continue;
+      }
+
+      auto bn_rm = constant_as<Tensor>(bn->namedInput("running_mean")).value();
+      auto bn_rv = constant_as<Tensor>(bn->namedInput("running_var")).value();
+      auto bn_eps = constant_as<double>(bn->namedInput("eps")).value();
+      auto linear_w = constant_as<Tensor>(linear->namedInput("weight")).value();
+
+      // implementation taken from torch/nn/utils/fusion.py
+      Tensor linear_b;
+      if (linear->namedInput("bias")->type() == NoneType::get()) {
+        at::ScalarType bias_dtype = bn_rm.scalar_type();
+        at::ScalarType weight_dtype = linear_w.scalar_type();
+        at::DeviceType weight_device = linear_w.device().type();
+        if (isBatchNorm1d && weight_device == at::kCUDA &&
```

Comment:
I'm not sure if we're expected to bypass this issue with bn{2,3}d (assuming correct usage). But for now, I've only checked for bn1d case here.